Refactor cli tools

lukpueh opened the pull request:
**Description of the changes being introduced by the pull request**:
This PR cleans up the command line interface code, by
-  adding consistent and sensible return codes,
- enhancing custom and autogenerated argparse help messages,
- removing obsolete library wrappers,
- fixing a small argparse bug in in_toto_record,
- unifying cli unittests

Read commit messages for more infos about the changes.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Remove library wrappers in command line tools

Remove wrapper function for corresponding library function that
passes through the library call in in_toto_{run, verify, mock}
command line tools.

This wrapper function does not serve a purpose and only requires
additional maintenance effort when the library api changes.

Commit:
Fix minor arg bug in_toto_record

Make subparser start/stop in in_toto_record cli tool required,
to fail early when none of the subcommands is specified.

Commit:
Replace parser.exit with parser.error

parser.error is a shortcut for parser.exit with return code 2,
indicating argument parsing errors.

Commit:
Enhance cli tool help messages

Add parser group for "required named arguments". To prevent
option based arguments from appearing in the "optional arguments"
help section even when required.

Simplify custom hardcoded help message and add metavars to
argument definitions, for enhanced auto help message.

Commit:
Add return code section to cli tool doc headers

Commit:
Add common tests module with CliTestCase class

CliTestCase is a subclass of unittest.TestCase and provides a
method that patches sys.argv (command line args) and expects a
SystemExit exception to be raised, which is handy for testing
the command line interfaces' argparsing methods.

Commit:
Adpot cli refactors in unit tests

All cli tools now return with sys.exit and a return code of one
of 0, 1 or 2. This commit adapts the cli tool calling methods to
test for that change, using the newly added CliTestCase helper
class.

Commit:
Fix line continuation in in_toto_verify

vladimir-v-diaz commented on in_toto/in_toto_keygen.py:
```diff
@@ -30,6 +30,12 @@
+  2 if an exception occurred during argument parsing
+  1 if an exception occurred
+  0 if no exception occurred
+
```
Apart from this docstring, are these return codes documented anywhere?  How do you inform users, who will not necessarily read the docstring, of these return codes?

lukpueh replied:
I don't think it's necessary, since this follows convention.

vladimir-v-diaz replied:
Which convention?

I think documenting return/status codes is standard practice.

Examples:
[HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)
[FTP return codes](https://en.wikipedia.org/wiki/List_of_FTP_server_return_codes)
[Pylint exit codes](https://docs.pylint.org/en/1.6.0/run.html#exit-codes)

Since you only really return 1-2 for exceptions, perhaps you can say that the script returns a non-zero value for any exception.

lukpueh replied:
Shell command exit codes follow the convention of returning non-zero in case of failure and zero in case of success.
However, documentation is provided in the form of above docstring.

vladimir-v-diaz commented on in_toto/in_toto_keygen.py:
```diff
@@ -30,6 +30,12 @@
 
   python in_toto_keygen.py -p bob_keys 2048
```
Do you plan to support additional algorithms with this CLI?  It might be good to state this somewhere.

lukpueh replied:
Not at the moment. This tool was initially developed to allow quick `securesystemslib` key pair creation from the command line with out having to open a Python interpreter and import modules. The intended use case was in-toto layout creation web tool, where it's reputability was challenged (https://github.com/in-toto/layout-web-tool/issues/17).

We can discuss if we want to extend this tool and maybe move it over to `securesystemslib` what do you think?

vladimir-v-diaz commented on in_toto/in_toto_keygen.py:
```diff
@@ -30,6 +30,12 @@
 
   python in_toto_keygen.py -p bob_keys 2048
```
If `in_toto_keygen.py` is added as a global script, you can simplify this example as:
```Bash
$ in_toto_keygen.py -p bob_keys 2048
```

lukpueh replied:
True, will change.

vladimir-v-diaz reviewed (commented):
My review of `in_toto_keygen.py`.

Some of the wrapped argument strings should use 4 spaces of indentation.

vladimir-v-diaz commented on in_toto/in_toto_mock.py:
```diff
@@ -42,56 +55,23 @@
 log = logging.getLogger("in_toto")
```
Is there a reason why you don't use the "a.b.c.." convention when specifying the logger name?
https://docs.python.org/2/library/logging.html#logging.getLogger

lukpueh replied:
Yes, there is and it is documented in [`in_toto.log`](https://github.com/in-toto/in-toto/blob/develop/in_toto/log.py#L35-L66) as suggested by the comment.

vladimir-v-diaz commented on in_toto/in_toto_mock.py:
```diff
@@ -31,8 +31,21 @@
   in-toto-mock --name write-code -- touch foo.py
```
I think you need to specify the `.py` extension with `in-toto-mock`. Actually, should that be `in_toto_mock.py`!?

Also, is the `--` by itself needed/correct in `-- touch foo.py`?

I was expecting it to be: `$ in-toto-mock.py --name write-code touch foo.py`

lukpueh replied:
Nope, [`in-toto-mock` works fine](https://github.com/in-toto/in-toto/blob/develop/setup.py#L80-L85) when in-toto is installed.

lukpueh replied:
Per *nix and [argparse](https://docs.python.org/dev/library/argparse.html#arguments-containing) convention is used to separate optional arguments from positional arguments.

vladimir-v-diaz commented on in_toto/in_toto_mock.py:
```diff
@@ -31,8 +31,21 @@
+  name:
+          A unique name to relate mock link metadata with a step defined
+          in the layout.
+  command:
```
The `command` argument in this docstring might be confusing to some users.  On the one hand, `--name` is an actual command-line argument, but `--command` is not.  That is, `$ in_toto_mock.py --name foo touch bar.log` is the expected format, where `touch bar.log` are positional arguments.

lukpueh replied:
Well technically, both are arguments, one *optional* (which I find confusing because it's actually required, see b19b17fdd4865896f70fb4f4ee8f8ebcb081d11d) and one *positional*.

Also, I think it's fine, because as you mentioned above, the user will probably not look at the docstring but rather at the command's help message, which should make the usage clear:
```
$ in-toto-mock -h
usage: in-toto-mock [-h] --name <unique step name> -- <command> [args]

Executes link command and records metadata

positional arguments:
  <command>             Link command to be executed with options and arguments

optional arguments:
  -h, --help            show this help message and exit

required named arguments:
  -n <unique step name>, --name <unique step name>
                        Unique name for link metadata
```

vladimir-v-diaz commented on in_toto/in_toto_mock.py:
```diff
@@ -100,7 +80,15 @@ def main():
   # TRY out in-toto-run, with fewer command lines args and max. user feedback.
```
I don't understand why this comment is here.  Should it go in the module docstring?

lukpueh replied:
Santiago requested to clarify the comment I had before an the PR that modified this line:
https://github.com/in-toto/in-toto/pull/183#discussion_r166432991.

The *"max. user feedback"* is the crucial part, that should explain why I set the log level to `INFO`.

vladimir-v-diaz replied:
Oh, it wasn't obvious to me how the comment relates to `log.setLevel(...)`.  Consider explicitly stating the connection, as you proposed in #183.

"in-toto-mock should not be used to secure the supply chain but only to try out in-toto-run, **hence it gives verbose user feedback**"

vladimir-v-diaz commented on in_toto/in_toto_mock.py:
```diff
@@ -100,7 +80,15 @@ def main():
+    in_toto.runlib.in_toto_mock(args.name, args.link_cmd)
+
+  except Exception as e:
+    log.error("in toto mock - {}".format(e))
```
Why is it written as `in toto mock` and not `in_toto_mock.py` or `in-toto mock`?

lukpueh replied:
Good question. I'll make uniform with `log.error`s in other command line tools.

vladimir-v-diaz commented on in_toto/in_toto_mock.py:
```diff
@@ -31,8 +31,21 @@
+
+
+<Return Codes>
+  2 if an exception occurred during argument parsing
```
Where is `2` returned in the code?

lukpueh replied:
By `ArgumentParser.error`, which can called by `argparse` internally when parsing errors, and I call it explicitly too (see 4ae540139e0831350ea49782ecd956a2ad3480d8).

vladimir-v-diaz reviewed (commented):
Here is my review of `in_toto_mock.py`.

vladimir-v-diaz commented on in_toto/in_toto_record.py:
```diff
@@ -76,6 +82,7 @@ def main():
   # The subparsers inherit the arguments from the parent parser
   parent_parser = argparse.ArgumentParser(add_help=False)
   subparsers = parser.add_subparsers(dest="command")
+  subparsers.required = True
```
Can you comment on why this `required = True` is... required?

lukpueh replied:
I hint at it in the related commit message (d74b274177c1ea0dc2ff71f8fceb9138e7d93af2), but I can add a comment too.

vladimir-v-diaz commented on in_toto/in_toto_run.py:
```diff
@@ -27,6 +27,40 @@
+  step_name:
+          A unique name to relate link metadata with a step defined in the
+          layout.
+  material_list:
```
Should `material_list` be `materials`?  Please verify the names of the other command-line options, like `products` instead of `product_list`.

lukpueh replied:
Yes, I can change this. We actually don't have a guideline for documenting arpgarse arguments. 

There are several ways:
 - an *abstract* name for the argument (that's what i did),
 - the argument name(s) the command line user has to use (e.g. `-n/--step-name`), which only works for positional arguments,
 - the variable where argparse stores argument (e.g. `step_name`),
 - a combination of above

What do you suggest?

lukpueh replied:
Another way would be to just copy-paste the output of the command's help message.
I think I'll do that, and I will probably extend the help messages a little bit.

vladimir-v-diaz replied:
I suggest something like the following:

```
<Arguments>
    -f/--foo (required argument):
        blah blah.
    -b/--bar (optional argument):
        bam bam.
    step-name (positional argument):
        Not a nickname, but a stepname.
```

You can also copy-paste the output of the command's help message.

vladimir-v-diaz commented on in_toto/in_toto_run.py:
```diff
@@ -40,106 +74,47 @@
-
-  in_toto_args = parser.add_argument_group("in-toto options")
+
+  parser.usage = ("%(prog)s <named arguments> [optional arguments]"
```
Is is better for the usage string to specify the required named arguments?

`(%s(prog)s --step-name foo --products ...)`

lukpueh replied:
That's what I had before, but it was messier to maintain. And the user can always use `--help` for a detailed description.

vladimir-v-diaz commented on in_toto/in_toto_run.py:
```diff
@@ -40,106 +74,47 @@
-      nargs='+', help="Files to record before link command execution")
+  parser.add_argument("-m", "--materials", type=str, required=False,
+      nargs='+', help="Files to record before link command execution",
+      metavar="<material path>")
```
Should you instead use `<material paths>` (note: plural) to indicate that more than one path can be given?

lukpueh replied:
No, given that metavar, argparse will create an accurate help message: 
```
-m <material path> [<material path> ...], --materials <material path> [<material path> ...]
```

vladimir-v-diaz commented on in_toto/in_toto_run.py:
```diff
@@ -40,106 +74,47 @@
       dest="record_streams", default=False, action="store_true")
 
-  in_toto_args.add_argument("-x", "--no-command",
+  parser.add_argument("-x", "--no-command",
```
I think this command-line argument is not documented in the module's docstring.

lukpueh replied:
Good eyes, I will add it.

vladimir-v-diaz reviewed (commented):
Here is my review of `in_toto_run.py`

vladimir-v-diaz commented on in_toto/in_toto_sign.py:
```diff
@@ -302,12 +302,12 @@ def main():
     if args.append:
       parser.print_help()
-      parser.exit(2, "wrong arguments: {}. Hence adding signatures to"
+      parser.error("wrong arguments: {}. Hence adding signatures to"
```
So the entire error message is:

```
("Link metadata is associated with a single functionary and is usually namespaced accordingly: '<name>.<keyid>.link'.")
Hence signing Link metadata with multiple keys is not allowed.
```

What if a functionary wants to use multiple keys?

vladimir-v-diaz replied:
For instance, one functionary wants the ability to sign with one of two keys under their control.  One key is stored on their laptop, and a different, and also trusted, key is on their Yubikey.

lukpueh replied:
This behavior  was already topic for discussion in https://github.com/in-toto/in-toto/pull/125 :)

The specification currently does not explicitly prohibit signing one link with multiple keys, however due to the naming convention `<related step name>.<SIGNING KEY PREFIX>.link` and the way final product verification will load links for steps, it does make sense and that's why this utility prohibits it.

Note that this doesn't mean that a functionary can't have multiple authorized keys to sign a link. All the public keys that a link can be signed with need to be embedded in the layout's `keys` section, and in to authorize them for a specific step, the keyid's must be in the step's `pubkeys` list.

vladimir-v-diaz reviewed (commented):
Comment for `in_toto_sign.py`

vladimir-v-diaz commented on in_toto/in_toto_verify.py:
```diff
@@ -16,12 +16,16 @@
   Provides a command line interface that wraps the verification of
   in_toto final product.
```
I've seen this a few times now.  Should `in_toto` be `in-toto`?

lukpueh replied:
Well, the project is called *in-toto*`, but due to Python naming restriction, the package is called `in_toto`. In above case it should be *in-toto*. I'll grep through the project and will change where necessary.

vladimir-v-diaz commented on in_toto/in_toto_verify.py:
```diff
@@ -36,6 +40,27 @@
+
+  layout keys:
+          List of path(s) to project owner public key(s), used to verify the
+          root layout's signature(s).
```
Can `in-toto-verify` verify any layout file?  I suppose it's referring to the root example above, but this comment makes it seem like only the root layout can be verified.

lukpueh replied:
Will change.

vladimir-v-diaz commented on in_toto/in_toto_verify.py:
```diff
@@ -45,104 +70,34 @@
-               "[--gpg-home <path to gpg keyring>]\n{0}"
-               "[--verbose | --quiet]\n\n"
-               .format(lpad))
+  named_args = parser.add_argument_group("required named arguments")
```
`<required arguments>` and `required_args` might be clearer.

lukpueh replied:
Ok, will change.

lukpueh replied:
Actually, I think I'll leave it as is. Because the `<required named arguments>` section doesn't list required positional arguments, which I would expect if it were called `<required arguments>`.
As for the variable name, `required_named_args` would probably be the best, but that seems a little long.

vladimir-v-diaz commented on in_toto/in_toto_verify.py:
```diff
@@ -160,10 +114,34 @@ def main():
     parser.print_help()
-    parser.exit(2, "wrong arguments: specify at least one of "
-      " `--layout-keys path [path ...]` or `--gpg id [id ...]`")
+    parser.error("wrong arguments: specify at least one of"
```
The Note: above states that passing both at the same time is possible, however, this parser.error implies otherwise.

lukpueh replied:
It says, at *"least one of keys [...] or gpg [...]"*, which IMHO doesn't imply that they are mutually exclusive.

vladimir-v-diaz commented on in_toto/in_toto_verify.py:
```diff
@@ -45,104 +70,34 @@
 
-  in_toto_args.add_argument("-l", "--layout", type=str, required=True,
-      help="Root layout to use for verification")
+  named_args.add_argument("-k", "--layout-keys", type=str,
```
This `add_argument()` call does not set required = True.  So is --layout-keys optional?  The `named_args` group indicates that they are required arguments.

lukpueh replied:
This is due to a shortcoming of `argparse`. There's no way of specifying a required (non mutual exclusive) group. If I'd make each argument required, the user couldn't specify only one of them, which is what I want.
I'll update the help message so that it's clear that only one of layout keys or gpg ids is okay.

vladimir-v-diaz commented on in_toto/in_toto_verify.py:
```diff
@@ -45,104 +70,34 @@
-  in_toto_args.add_argument("-l", "--layout", type=str, required=True,
-      help="Root layout to use for verification")
+  named_args.add_argument("-k", "--layout-keys", type=str,
+      nargs="+", help="Key(s) to verify root layout signature",
```
Is there only a root layout?  In other words, are there no general layouts?  Are signatures stored in something called a root layout?

lukpueh replied:
We have not exact definition of what a *root layout* is as opposed to a *layout*.

In my understanding, the root layout of a project specifies the entire supply chain for that project, however steps of the supply chain can be delegated to sublayouts (with arbitrary nesting). As a consequence, a root layout of a project can itself be a sublayout of another project.

In order to verify the supply chain of the final product of a project, you need to pass that project's root layout to `in-toto-verify`.

vladimir-v-diaz reviewed (commented):
My review of `in-toto-verify.py`.

vladimir-v-diaz commented on tests/common.py:
```diff
@@ -0,0 +1,79 @@
+
+<Purpose>
+  Common code for in-toto unittests, import like so:
+  `import tests.common`
```
It might help if you explained how this module should be used.

Namely, that test cases inherit from CliTestCase:
```
class MyTestCase(tests.common.CliTestCase):
```

Do you also want to mention the provided convenience/mocking methods?

lukpueh replied:
I think that the class and method docstrings of `CLITestCase` describes its usage well enough.

vladimir-v-diaz reviewed (commented):
My review of `tests/common.py`.

vladimir-v-diaz commented:
@lukpueh 
Okay, finished reviewing all of #184.

Commit:
Add missing shebang to cli tools

Commit:
Unify error messages in cli tools

Commit:
Replace some occurrences of in_toto with in-toto

"in-toto" is the name of the project, whereas "in_toto" is the
name of the Python package (due to naming restrictions).
This commit replaces some occurrences of in_toto, where the
project is meant.

Commit:
Rephrase in-toto-mock log level comment

Commit:
Add comment for subparsers.required workaround

The workaround is taken from
https://stackoverflow.com/questions/18282403/argparse-with-required-subcommands
http://bugs.python.org/issue9253#msg186387

Commit:
Add and refactor argparse metavars

Add metavars to in_toto_sign and refactor existing metavars, that
is make them shorter, for a less convoluted help page.
It seems better to provide additional information about a given
argument in its help message instead of its metavar name.

Commit:
Enhance in-toto-run help message

- Rephrase tool description
- Add tool epilog with examples
- Rephrase argument help texts

- Copy paste output of in-toto-run -h to doc header

Commit:
Enhance in-toto-verify help message

- Rephrase tool description
- Add tool epilog with examples
- Rephrase argument help texts

- Copy paste output of in-toto-verify -h to doc header

Commit:
Enhance in-toto-mock help message

- Rephrase tool description
- Add tool epilog with examples
- Rephrase argument help texts

- Copy paste output of in-toto-mock -h to doc header

Commit:
Enhance in-toto-record help message

- Rephrase tool description
- Add tool epilog with examples
- Rephrase argument help texts

- Copy paste output of in-toto-record -h to doc header including
  help messages from subcommands start and stop

Commit:
Enhance in-toto-sign help message

- Rephrase tool description
- Add tool epilog with examples
- Rephrase argument help texts

- Copy paste output of in-toto-sign -h to doc header.

Commit:
Update usage in doc header of in_toto_keygen

lukpueh commented:
Thanks for the detailed review, @vladimir-v-diaz. 
I addressed all your comments and also improved the help messages for all tools (except for keygen) massively. Please let me know what you think.

Commit:
Add spacing in cli tool epilogs and fix typos

Commit:
Rename --record-streams short option

--record-streams used to be called --record-byproducts. The long
option name was changed in 2149e0c155d6d6310231fe0a5f8b6b648300b441
but not the short option.
This commit renames -b to -s accordingly.

lukpueh commented:
I gave the help texts another pass and fixed some spacing and typos. I'd say I addressed all of @vladimir-v-diaz's comments. Thanks for reviewing! Merging...
